src/esi/Sequence.cc

```diff
@@ -88,12 +91,13 @@ esiSequence::render(ESISegment::Pointer output)
 
     for (size_t i = 0; i < processedcount; ++i) {
         elements[i]->render(output);
-        elements.setNULL(i,i+1);
+        FinishAnElement(elements[i], i);
         /* FIXME: pass a ESISegment ** ? */
         output = output->tail();
     }
 
-    elements.pop_front (processedcount);
+    // prune completed (or otherwise unset) elements
+    std::remove_if(elements.begin(), elements.end(), [](const ESIElement::Pointer &e) { return !e; });
```

Comment:
Yes there is rather a lot of access by operator[] for elements. But more importantly there is some possible access by pointers which in std::deque is undefined behaviour. Once that latter is cleaned up something better than vector might be used. I am leaving it out of scope for now.

I went with remove_if rather than erase() of the prefix range since there is other processing code which sets randomly accessed entries to nullptr. So this call prunes all those as well resulting in a smaller set for future processing at expense of extra work initially. If you think we should be using erase(begin(), N) instead I can change it, but is that really worthwhile?

Reply:
> If you think we should be using erase(begin(), N) instead I can change it, but is that really worthwhile?

Unfortunately, I do not have enough information to know whether your optimization is worthwhile or harmful. In other words, I do not know whether changing the old behavior will result in slower or faster code. Perhaps you do know that and just have chosen not to share that information. 

My logic is simple: Since the changed code uses a significantly more expensive operation (multiple array shifts instead of one), if you are not sure that you are making things overall faster, then do not change the old behavior in this PR. In other words, use erase() or an equivalent single-shift operation.

Reply:
okay, done.